src/v/raft/tests/group_configuration_tests.cc

```diff
@@ -7,114 +7,427 @@
+        ","));
+    return o;
+}
+
+class ConfigurationCancellationTest
+  : public TestWithParam<configuration_cancel_test_params> {};
+
+/**
+ * This test verifies if cancelling the configuration change twice at any point
+ * will lead to the cancellation being reverted.
+ * example:
+ * replicas are updated from set A to set B
+ *  1. A->B
+ * then a reconfiguration is cancelled
+ *  2. B->A
+ * then the reconfiguration is cancelled again
+ *  3. A->B
+ * Finally the replica set must be equal to B
+ */
+TEST_P(ConfigurationCancellationTest, TestEvenNumberOfCancellations) {
+    const auto params = GetParam();
+
+    const std::vector<raft::vnode> original_replicas
+      = params.initial_replica_set;
+    const std::vector<raft::vnode> target_replicas = params.target_replica_set;
+    const auto to_add = diff(target_replicas, original_replicas);
+    const auto to_remove = diff(original_replicas, target_replicas);
+
+    raft::group_configuration test_cfg = raft::group_configuration(
+      original_replicas, model::revision_id(0));
+    test_cfg.set_version(raft::group_configuration::v_7);
+
+    // trigger reconfiguration
+    test_cfg.replace(target_replicas, model::revision_id{0}, std::nullopt);
+
+    ASSERT_THAT(
+      test_cfg.get_configuration_update()->replicas_to_add,
+      ElementsAreArray(to_add));
+    ASSERT_THAT(
+      test_cfg.get_configuration_update()->replicas_to_remove,
+      ElementsAreArray(to_remove));
+
+    // CASE 1. Cancel right after change was requested
+
+    // cancel configuration change, goes straight to simple state as learner can
+    // be removed immediately
+    auto cfg_1 = test_cfg;
+    cfg_1.cancel_configuration_change(model::revision_id{0});
+    // check if reconfiguration is finished by optimizations
+    if (cfg_1.get_state() == raft::configuration_state::simple) {
+        ASSERT_EQ(cfg_1.current_config().voters, original_replicas);
+        ASSERT_TRUE(cfg_1.current_config().learners.empty());
+    } else {
+        cfg_1.cancel_configuration_change(model::revision_id{0});
+        transition_configuration_update(cfg_1);
+        ASSERT_EQ(cfg_1.current_config().voters, target_replicas);
+        ASSERT_EQ(cfg_1.get_state(), raft::configuration_state::simple);
+        ASSERT_TRUE(cfg_1.current_config().learners.empty());
+    }
+    // CASE 2. Cancel after learners promotion
+    for (size_t cancel_after = 1; cancel_after <= to_add.size();
+         ++cancel_after) {
+        // create a copy of the configuration for each round
+        auto cfg = test_cfg;
+        for (size_t i = 0; i < cancel_after; ++i) {
+            cfg.promote_to_voter(to_add[i]);
+        }
+        // update the original configuration for the next step
+        if (cancel_after == to_add.size()) {
+            test_cfg = cfg;
+        }
+        cfg.cancel_configuration_change(model::revision_id{0});
+        cfg.cancel_configuration_change(model::revision_id{0});
```

Comment:
I wonder if we can let the configuration change run in the cancelled direction for some steps instead of cancelling second time right away? I.e. have a "step" function (promoting learners etc.) and let it run an arbitrary number of steps in 1 direction, then cancel, do arbitrary number of steps, cancel again, then step till it the cfg is simple. Will be nicely uniform.

Reply:
good point, i will add another test case like this